File: src/main/core/profiles.js
'use strict';

const { JsonStore } = require('./store');

/**
 * Профили модов — как в Vortex, r2modman и CurseForge (1.10).
 *
 * Профиль — это просто запомненный список «что включено». Моды не
 * копируются: включение и выключение в ModHub и так сделано переносом
 * папки (см. registry.js), поэтому переключить профиль — значит включить
 * одни папки и выключить другие. Быстро, обратимо и без дублей на диске.
 */

const MAX_PROFILES = 30;

function cleanName(value) {
  const name = String(value ?? '')
    .replace(/[\u0000-\u001f]/g, '')
    .trim()
    .slice(0, 40);
  // Имя — ключ объекта в JSON: служебные имена объектов не пускаем.
  return ['__proto__', 'constructor', 'prototype'].includes(name) ? '' : name;
}

class Profiles {
  /** @param {{file: string, now?: () => number}} options */
  constructor({ file, now = () => Date.now() }) {
    this.store = new JsonStore(file, { games: {}, active: {} });
    this.now = now;
  }

  _game(gameId) {
    if (!this.store.data.games[gameId]) this.store.data.games[gameId] = {};
    return this.store.data.games[gameId];
  }

  /** Профили игры, новые — первыми. */
  list(gameId) {
    const all = this.store.data.games[gameId] ?? {};
    return Object.entries(all)
      .map(([name, p]) => ({
        name,
        createdAt: p.createdAt,
        updatedAt: p.updatedAt ?? p.createdAt,
        count: p.enabled.length,
        active: this.store.data.active[gameId] === name,
      }))
      .sort((a, b) => String(b.updatedAt).localeCompare(String(a.updatedAt)));
  }

  get(gameId, name) {
    return this.store.data.games[gameId]?.[name] ?? null;
  }

  /**
   * Запомнить то, что сейчас включено, под именем.
   * @param {string} gameId
   * @param {string} rawName
   * @param {object[]} records записи реестра
   */
  save(gameId, rawName, records) {
    const name = cleanName(rawName);
    if (!name) throw Object.assign(new Error('empty profile name'), { code: 'PROFILE_NAME' });
    const game = this._game(gameId);
    if (!game[name] && Object.keys(game).length >= MAX_PROFILES) {
      throw Object.assign(new Error('too many profiles'), { code: 'PROFILE_LIMIT' });
    }
    const at = new Date(this.now()).toISOString();
    const enabled = (records ?? []).filter((r) => r.enabled !== false && !r.missing).map((r) => r.id);
    game[name] = { createdAt: game[name]?.createdAt ?? at, updatedAt: at, enabled };
    this.store.data.active[gameId] = name;
    this.store.save();
    return this.get(gameId, name);
  }

  /**
   * Включить моды профиля и выключить остальные.
   * @param {string} gameId
   * @param {string} name
   * @param {import('./registry').ModRegistry} registry
   * @returns {{enabled: number, disabled: number, missing: string[]}}
   */
  apply(gameId, name, registry) {
    const profile = this.get(gameId, name);
    if (!profile) throw Object.assign(new Error('no such profile'), { code: 'PROFILE_MISSING' });
    const wanted = new Set(profile.enabled);
    let enabled = 0;
    let disabled = 0;
    for (const mod of registry.list()) {
      const on = wanted.has(mod.id);
      if (mod.enabled === on) continue;
      registry.setEnabled(mod.id, on);
      if (on) enabled += 1;
      else disabled += 1;
    }
    // Моды, которые были в профиле, но с тех пор удалены.
    const missing = profile.enabled.filter((id) => !registry.has(id));
    this.store.data.active[gameId] = name;
    this.store.save();
    return { enabled, disabled, missing };
  }

  rename(gameId, from, rawTo) {
    const to = cleanName(rawTo);
    const game = this._game(gameId);
    if (!game[from] || !to) return false;
    if (to !== from) {
      game[to] = game[from];
      delete game[from];
      if (this.store.data.active[gameId] === from) this.store.data.active[gameId] = to;
      this.store.save();
    }
    return true;
  }

  remove(gameId, name) {
    const game = this._game(gameId);
    if (!game[name]) return false;
    delete game[name];
    if (this.store.data.active[gameId] === name) delete this.store.data.active[gameId];
    this.store.save();
    return true;
  }
}

module.exports = { Profiles, cleanName };

File: desktop/ModLaunch/Features/Profiles.cs
using System.Text.Json.Nodes;
using ModLaunch.Core;
using ModLaunch.Mods;

namespace ModLaunch.Features;

public sealed record ProfileInfo(string Name, DateTime? UpdatedAt, int Count, bool Active);

/// <summary>Профили: какие моды включены. Хранятся в profiles.json, как в 3.x.</summary>
public static class Profiles
{
    const int Max = 30;
    static readonly JsonFile File = new(Path.Combine(Paths.DataDir, "profiles.json"), () => new JsonObject { ["games"] = new JsonObject(), ["active"] = new JsonObject() });

    static JsonObject Game(string gameId) => File.Data.Obj("games").Obj(gameId);
    static JsonObject Active => File.Data.Obj("active");

    public static string Clean(string? value)
    {
        var name = new string((value ?? "").Where(c => c >= ' ').ToArray()).Trim();
        return name.Length > 40 ? name[..40] : name;
    }

    public static List<ProfileInfo> List(string gameId) =>
        Game(gameId).Select(kv => new ProfileInfo(
                kv.Key,
                DateTime.TryParse(kv.Value.Str("updatedAt") ?? kv.Value.Str("createdAt"), out var d) ? d.ToUniversalTime() : null,
                kv.Value.Arr("enabled").Count,
                Active.Str(gameId) == kv.Key))
            .OrderByDescending(p => p.UpdatedAt).ToList();

    public static void Save(string gameId, string rawName, ModRegistry registry)
    {
        var name = Clean(rawName);
        if (name == "") throw new InvalidOperationException(I18n.T("err.PROFILE_NAME"));
        var game = Game(gameId);
        if (!game.ContainsKey(name) && game.Count >= Max) throw new InvalidOperationException(I18n.T("err.PROFILE_LIMIT"));
        registry.Reconcile();
        var at = DateTime.UtcNow.ToString("o");
        var enabled = registry.List().Where(r => r.Bool("enabled", true) && !r.Bool("missing")).Select(r => (JsonNode)r.Str("id")!).ToArray();
        game[name] = new JsonObject { ["createdAt"] = game[name].Str("createdAt") ?? at, ["updatedAt"] = at, ["enabled"] = new JsonArray(enabled) };
        Active[gameId] = name;
        File.Save();
    }

    public static (int On, int Off, int Missing) Apply(string gameId, string name, ModRegistry registry)
    {
        if (Game(gameId)[name] is not JsonObject profile) throw new InvalidOperationException(I18n.T("err.PROFILE_MISSING"));
        registry.Reconcile();
        var wanted = profile.Arr("enabled").Select(x => x?.ToString()).OfType<string>().ToHashSet();
        int on = 0, off = 0;
        foreach (var mod in registry.List())
        {
            var id = mod.Str("id")!;
            var want = wanted.Contains(id);
            if (mod.Bool("enabled", true) == want) continue;
            registry.SetEnabled(id, want);
            if (want) on++; else off++;
        }
        Active[gameId] = name;
        File.Save();
        return (on, off, wanted.Count(id => !registry.Has(id)));
    }

    public static void Rename(string gameId, string from, string rawTo)
    {
        var to = Clean(rawTo);
        var game = Game(gameId);
        if (to == "" || to == from || game[from] is not JsonNode node) return;
        game.Remove(from);
        game[to] = node;
        if (Active.Str(gameId) == from) Active[gameId] = to;
        File.Save();
    }

    public static void Remove(string gameId, string name)
    {
        Game(gameId).Remove(name);
        if (Active.Str(gameId) == name) Active.Remove(gameId);
        File.Save();
    }
}
